// fable-react/umd/react-native-onesignal/Fable.Helpers.ReactNativeOneSignal.js
(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(["exports", "react-native-onesignal"], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require("react-native-onesignal"));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.reactNativeOnesignal);
        global.FableHelpersReactNativeOneSignal = mod.exports;
    }
})(this, function (exports, _reactNativeOnesignal) {
    "use strict";

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.OneSignal = undefined;

    var _reactNativeOnesignal2 = _interopRequireDefault(_reactNativeOnesignal);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    var OneSignal = exports.OneSignal = function (__exports) {
        var configure = __exports.configure = function (props) {
            _reactNativeOnesignal2.default.configure(props);
        };

        var enableInAppAlertNotification = __exports.enableInAppAlertNotification = function () {
            _reactNativeOnesignal2.default.enableInAppAlertNotification(true);
        };

        var disableInAppAlertNotification = __exports.disableInAppAlertNotification = function () {
            _reactNativeOnesignal2.default.enableInAppAlertNotification(false);
        };

        var enableNotificationsWhenActive = __exports.enableNotificationsWhenActive = function () {
            _reactNativeOnesignal2.default.enableNotificationsWhenActive(true);
        };

        var disableNotificationsWhenActive = __exports.disableNotificationsWhenActive = function () {
            _reactNativeOnesignal2.default.enableNotificationsWhenActive(false);
        };

        var enableSound = __exports.enableSound = function () {
            _reactNativeOnesignal2.default.enableSound(true);
        };

        var disableSound = __exports.disableSound = function () {
            _reactNativeOnesignal2.default.enableSound(false);
        };

        var enableVibrate = __exports.enableVibrate = function () {
            _reactNativeOnesignal2.default.enableVibrate(true);
        };

        var disableVibrate = __exports.disableVibrate = function () {
            _reactNativeOnesignal2.default.enableVibrate(false);
        };

        var enableSubscriptions = __exports.enableSubscriptions = function () {
            _reactNativeOnesignal2.default.setSubscription(true);
        };

        var disableSubscriptions = __exports.disableSubscriptions = function () {
            _reactNativeOnesignal2.default.setSubscription(false);
        };

        var promptLocation = __exports.promptLocation = function () {
            _reactNativeOnesignal2.default.promptLocation(null);
        };

        var clearOneSignalNotifications = __exports.clearOneSignalNotifications = function () {
            _reactNativeOnesignal2.default.clearOneSignalNotifications(null);
        };

        var cancelNotification = __exports.cancelNotification = function (id) {
            _reactNativeOnesignal2.default.cancelNotification(id);
        };

        var postP2PNotification = __exports.postP2PNotification = function (message, data, userId) {
            _reactNativeOnesignal2.default.postNotification(message, data, userId);
        };

        return __exports;
    }({});
});
