// fable-react/react-native-onesignal/Fable.Helpers.ReactNativeOneSignal.js
import react_native_onesignal from "react-native-onesignal";
export var OneSignal = function (__exports) {
    var configure = __exports.configure = function (props) {
        react_native_onesignal.configure(props);
    };

    var enableInAppAlertNotification = __exports.enableInAppAlertNotification = function () {
        react_native_onesignal.enableInAppAlertNotification(true);
    };

    var disableInAppAlertNotification = __exports.disableInAppAlertNotification = function () {
        react_native_onesignal.enableInAppAlertNotification(false);
    };

    var enableNotificationsWhenActive = __exports.enableNotificationsWhenActive = function () {
        react_native_onesignal.enableNotificationsWhenActive(true);
    };

    var disableNotificationsWhenActive = __exports.disableNotificationsWhenActive = function () {
        react_native_onesignal.enableNotificationsWhenActive(false);
    };

    var enableSound = __exports.enableSound = function () {
        react_native_onesignal.enableSound(true);
    };

    var disableSound = __exports.disableSound = function () {
        react_native_onesignal.enableSound(false);
    };

    var enableVibrate = __exports.enableVibrate = function () {
        react_native_onesignal.enableVibrate(true);
    };

    var disableVibrate = __exports.disableVibrate = function () {
        react_native_onesignal.enableVibrate(false);
    };

    var enableSubscriptions = __exports.enableSubscriptions = function () {
        react_native_onesignal.setSubscription(true);
    };

    var disableSubscriptions = __exports.disableSubscriptions = function () {
        react_native_onesignal.setSubscription(false);
    };

    var promptLocation = __exports.promptLocation = function () {
        react_native_onesignal.promptLocation(null);
    };

    var clearOneSignalNotifications = __exports.clearOneSignalNotifications = function () {
        react_native_onesignal.clearOneSignalNotifications(null);
    };

    var cancelNotification = __exports.cancelNotification = function (id) {
        react_native_onesignal.cancelNotification(id);
    };

    var postP2PNotification = __exports.postP2PNotification = function (message, data, userId) {
        react_native_onesignal.postNotification(message, data, userId);
    };

    return __exports;
}({});
