// fable-react/react-native-push-notification/Fable.Helpers.ReactNativePushNotification.js
import react_native_push_notification from "react-native-push-notification";
export var PushNotification = function (__exports) {
    var configure = __exports.configure = function (props) {
        react_native_push_notification.configure(props);
    };

    var setApplicationIconBadgeNumber = __exports.setApplicationIconBadgeNumber = function (number) {
        react_native_push_notification.setApplicationIconBadgeNumber(number);
    };

    var localNotification = __exports.localNotification = function (props, message) {
        react_native_push_notification.localNotification(Object.assign({
            message: message
        }, props));
    };

    var localNotificationSchedule = __exports.localNotificationSchedule = function (props, date, message) {
        react_native_push_notification.localNotificationSchedule(Object.assign({
            message: message,
            date: date
        }, props));
    };

    var cancelLocalNotifications = __exports.cancelLocalNotifications = function (details) {
        react_native_push_notification.cancelLocalNotifications(details);
    };

    var cancelAllLocalNotifications = __exports.cancelAllLocalNotifications = function () {
        react_native_push_notification.cancelAllLocalNotifications(null);
    };

    return __exports;
}({});
